README.md
# knot-cloud-sdk-js-authenticator

KNoT Cloud authenticator service JavaScript library.

## Quickstart

### Install

```console
npm install --save @cesarbr/knot-cloud-sdk-js-authenticator
```

### Run

`KNoTCloudAuthenticator` connects to &lt;protocol&gt;://&lt;hostname&gt;:&lt;port&gt; using email and password as credentials. Replace this address with your authenticator instance and the credentials with valid ones.

```javascript
const KNoTCloudAuthenticator = require("@cesarbr/knot-cloud-sdk-js-authenticator");

const client = new KNoTCloudAuthenticator({
  protocol: "https",
  hostname: "api.knot.cloud",
  port: 443,
});

async function main() {
  try {
    const user = await client.createUser(
      "awesome@email.com",
      "strong-password"
    );
    console.log(user);
    const token = client.createToken("awesome@email.com", "strong-password");
    console.log(token);
  } catch (err) {
    console.error(err);
  }
}
main();
```

## Methods

### Constructor(options)

Create a client object that will connect to a KNoT Cloud protocol authenticator instance.

#### Arguments

- `config` **Object** JSON object with request details.
  - `protocol` **String** (Optional) Either `'http'` or `'https'`. Default: `'https'`.
  - `hostname` **String** KNoT Cloud authenticator instance hostname. Default `'api.knot.cloud'`.
  - `port` **Number** (Optional) KNoT Cloud authenticator instance port. Default: 80/443.

#### Example

```javascript
const KNoTCloudAuthenticator = require("@cesarbr/knot-cloud-sdk-js-authenticator");

const client = new KNoTCloudAuthenticator({
  protocol: "https",
  hostname: "api.knot.cloud",
  port: 443,
});
```

### createUser(email, password) &lt;Object&gt;

Creates a new user.

#### Arguments

- `email` **String** User email.
- `password` **String** User password in plain text.

#### Result

- `user` **Object** JSON object containing the created user's credentials.

#### Example

```javascript
const KNoTCloudAuthenticator = require("@cesarbr/knot-cloud-sdk-js-authenticator");

const client = new KNoTCloudAuthenticator({
  protocol: "https",
  hostname: "api.knot.cloud",
  port: 443,
});

async function main() {
  try {
    const user = await client.createUser(
      "awesome@email.com",
      "strong-password"
    );
  } catch (err) {
    console.error(err);
  }
}
main();

// {
//  email: 'awesome@email.com',
//  password: 'strong-password',
// }
```

### createToken(email, password): &lt;Object&gt;

Generates a new access token for the user.

#### Arguments

- `email` **String** User email.
- `password` **String** User password, in plain text.

#### Result

- `token` **Object** JSON object containing the user's new access token.

#### Example

```javascript
const KNoTCloudAuthenticator = require("@cesarbr/knot-cloud-sdk-js-authenticator");

const client = new KNoTCloudAuthenticator({
  protocol: "https",
  hostname: "api.knot.cloud",
  port: 443,
});

async function main() {
  try {
    const token = await client.createToken(
      "awesome@email.com",
      "strong-password"
    );
    console.log(token);
  } catch (err) {
    console.error(err);
  }
}
main();

// {
//  token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1ODcxMjkzNzIsImlhdCI6MTU4NzA5MzM3MiwiaXNzIjoibWFpbmZsdXguYXV0aG4iLCJzdWIiOiJkYWRhdmR2YkBrbm90LmNvbSIsInR5cGUiOjB9._lbRa2fzI_CvEorbEACVAf2UnHvkiCOORY55wCWUGAs'
// }
```


CHANGELOG.md
# [2.1.2](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/compare/v2.1.1...v2.1.2)

 ### Bug Fixes

  - Fix vulnerability issues

# [2.1.1](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/compare/v2.1.0...v2.1.1)
 
 ### Bug Fixes

- Add pathname parameter to HTTP component.

# [2.1.0](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/compare/v2.0.1...v2.1.0)

 ### Features

- Update `createToken` to support 'app' token creation with a configurable duration.

# [2.0.1](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/compare/v2.0.0...v2.0.1)

 ### Bug Fixes

- Return status code when handling error.

# [2.0.0](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/compare/v1.0.0...v2.0.0)

 ### Features

- Update methods to communicate with the new users authentication service as can be seen [here](https://github.com/CESARBR/knot-babeltower/blob/master/docs/swagger.yaml).
    - Compatibilize create user operation.
    - Add create token operation.
    - Remove forget password and reset password operations.

# [1.0.0](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/compare/42518d4...v1.0.0)

 ### Features

- Methods to communicate with [knot-cloud-authenticator](https://github.com/CESARBR/knot-cloud-authenticator/).
  - Create user
  - Authenticate user
  - Forgot password
  - Reset password


src/Client.test.js
import HTTP, * as httpMocks from './network/HTTP';
import Client from './Client';

jest.mock('./network/HTTP');

const email = 'knot@knot.com';
const password = '123qwe123qwe';

describe('SDK Authenticator', () => {
  beforeEach(() => {
    httpMocks.mockPost.mockClear();
  });

  test('should create a new user when connection is ok', async () => {
    const http = new HTTP();
    const client = new Client(http);
    await client.createUser(email, password);
    expect(httpMocks.mockPost).toHaveBeenCalled();
  });

  test('should fail to create a new user when something goes wrong', async () => {
    const postErr = 'error in post request';
    const http = new HTTP({ postErr });
    const client = new Client(http);
    let error;
    try {
      await client.createUser(email, password);
    } catch (err) {
      error = err.message;
    }
    expect(error).toBe(postErr);
  });
});


test/unit/client.test.js
/* global describe,test,beforeEach,expect,jest */
import axios from 'axios';
import Client from 'Client';

jest.mock('axios');

describe('Client', () => {
  let client;

  beforeEach(() => {
    client = new Client({ protocol: 'http', hostname: 'test' });
  });

  describe('constructor', () => {
    test('should set protocol https and port 443 as default', async () => {
      const instance = new Client({ hostname: 'test' });
      const response = { data: { uuid: 'uuid-mocked', token: 'token-mocked' } };
      axios.post.mockResolvedValueOnce(response);

      await instance.createUser('test@test.com', '123qwe!@#QWE');

      const calledUri = axios.post.mock.calls[0][0];
      expect(calledUri).toBe('https://test:443/users');
    });

    test('should throw missing hostname when constructed without hostname', () => {
      expect(() => new Client({}))
        .toThrow();
    });

    test('should throw unexpected protocol when constructed without http nor https', () => {
      expect(() => new Client({ protocol: 'invalid-protocol', hostname: 'test' }))
        .toThrow();
    });
  });

  describe('create user', () => {
    test('should call post', async () => {
      const user = { email: 'test@test.com', password: '123qwe!@#QWE' };
      const resp = { data: { uuid: 'uuid-mocked', token: 'token-mocked' } };

      axios.post.mockResolvedValueOnce(resp);

      await client.createUser(user.email, user.password);

      return expect(axios.post).toBeCalled();
    });

    test("uri path should be 'http://test:443/users'", async () => {
      expect(axios.post).lastCalledWith('http://test:443/users', { email: 'test@test.com', password: '123qwe!@#QWE' });
    });

    test('should create user', async () => {
      const user = { email: 'test@test.com', password: '123qwe!@#QWE' };
      const resp = { data: { uuid: 'uuid-mocked', token: 'token-mocked' } };

      axios.post.mockResolvedValueOnce(resp);

      const data = await client.createUser(user.email, user.password);

      return expect(data).toEqual(resp.data);
    });

    test('should throw error when could not create user', async () => {
      const user = { email: 'test@test.com', password: '123qwe!@#QWE' };

      axios.post.mockRejectedValueOnce(new Error('unknown error'));

      await expect(client.createUser(user.email, user.password)).rejects.toThrow();
    });
  });

  describe('authenticate user', () => {
    test('should call post', async () => {
      const user = { email: 'test@test.com', password: '123qwe!@#QWE' };
      const resp = { data: { uuid: 'uuid-mocked', token: 'token-mocked' } };

      axios.post.mockResolvedValueOnce(resp);

      await client.authUser(user.email, user.password);

      return expect(axios.post).toBeCalled();
    });

    test("uri path should be 'http://test:443/auth'", async () => {
      expect(axios.post).lastCalledWith('http://test:443/auth', { email: 'test@test.com', password: '123qwe!@#QWE' });
    });

    test('should auth user', async () => {
      const user = { email: 'test@test.com', password: '123qwe!@#QWE' };
      const resp = { data: { uuid: 'uuid-mocked', token: 'token-mocked' } };

      axios.post.mockResolvedValueOnce(resp);

      const data = await client.authUser(user.email, user.password);

      return expect(data).toEqual(resp.data);
    });

    test('should throw error when could not authenticate an user', async () => {
      const user = { email: 'test@test.com', password: '123qwe!@#QWE' };

      axios.post.mockRejectedValueOnce(new Error('unknown error'));

      await expect(client.authUser(user.email, user.password)).rejects.toThrow();
    });
  });

  describe('forgot user', () => {
    test('should call post', async () => {
      const user = { email: 'test@test.com' };

      await client.forgotPassword(user.email);

      return expect(axios.post).toBeCalled();
    });

    test("uri path should be 'http://test:443/forgot'", async () => {
      expect(axios.post).lastCalledWith('http://test:443/forgot', { email: 'test@test.com' });
    });

    test('should return undefined', async () => {
      const user = { email: 'test@test.com' };

      axios.post.mockResolvedValueOnce();

      const data = await client.forgotPassword(user.email);

      return expect(data).toBeUndefined();
    });

    test('should throw error when internal server error ', async () => {
      const user = { email: 'test@test.com' };

      axios.post.mockRejectedValueOnce(new Error('unknown error'));

      await expect(client.forgotPassword(user.email)).rejects.toThrow();
    });
  });

  describe('reset password', () => {
    test('should call post', async () => {
      const user = { email: 'test@test.com', password: 'new-password' };
      const token = 'mocked-token';

      axios.post.mockResolvedValue();

      await client.resetPassword(user.email, token, user.password);

      return expect(axios.post).toBeCalled();
    });

    test("uri path should be 'http://test:443/reset'", async () => {
      expect(axios.post).lastCalledWith('http://test:443/reset',
        { email: 'test@test.com', token: 'mocked-token', password: 'new-password' });
    });

    test('should return undefined', async () => {
      const user = { email: 'test@test.com', password: 'new-password' };
      const token = 'mocked-token';

      axios.post.mockResolvedValueOnce();

      const data = await client.resetPassword(user.email, token, user.password);

      return expect(data).toBeUndefined();
    });


    test('should throw error when internal server error', async () => {
      const user = { email: 'test@test.com', password: 'new-password' };
      const token = 'mocked-token';

      axios.post.mockRejectedValueOnce(new Error('unknown error'));

      await expect(client.resetPassword(user.email, token, user.password)).rejects.toThrow();
    });
  });
});


rollup.config.js
/* eslint-disable import/no-extraneous-dependencies */
import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import resolve from 'rollup-plugin-node-resolve';
import builtins from 'rollup-plugin-node-builtins';
import globals from 'rollup-plugin-node-globals';
import autoExternal from 'rollup-plugin-auto-external';

import pkg from './package.json';

export default [
  {
    input: 'src/index.js',
    output: [
      {
        file: pkg.browser,
        format: 'umd',
        sourceMap: true,
        name: 'KNoTAuthenticator',
      },
    ],
    plugins: [
      resolve({
        browser: true,
        preferBuiltins: true,
      }),
      commonjs(),
      babel({
        exclude: ['node_modules/**'],
        externalHelpers: false,
        runtimeHelpers: true,
      }),
      globals(),
      builtins(),
    ],
  },
  {
    input: 'src/index.js',
    output: [
      {
        file: pkg.main,
        format: 'cjs',
        sourceMap: true,
      },
      {
        file: pkg.module,
        format: 'es',
        sourceMap: true,
      },
    ],
    plugins: [
      autoExternal({
        builtins: '6.0.0',
      }),
      babel({
        exclude: ['node_modules/**'],
        externalHelpers: false,
        runtimeHelpers: true,
      }),
      resolve(),
      commonjs(),
    ],
  },
];


src/network/HTTP.js
import axios from 'axios';

export default class HTTP {
  constructor({
    protocol = 'https',
    hostname = 'api.knot.cloud',
    port = (protocol === 'https') ? 443 : 80,
    pathname = '',
  }) {
    if ((protocol !== 'http') && (protocol !== 'https')) {
      throw new Error('Invalid protocol: must be either \'https\' or \'http\'');
    }
    if (!hostname) {
      throw new Error('Required field missing: \'hostname\'');
    }

    this.baseUrl = `${protocol}://${hostname}:${port}${pathname}`;
    this.header = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    };
  }

  async post(path, data) {
    const config = {
      url: `${this.baseUrl}/${path}`,
      method: 'POST',
      headers: this.header,
      data: data && JSON.stringify(data),
    };

    return axios(config)
      .then(res => Promise.resolve(res.data || JSON.parse(res.config.data)))
      .catch(err => this.handleError(err));
  }

  async handleError(r) {
    /*
    * Reject with a detailed error if the response was correctly received.
    * Otherwise, reject with an unexpected error.
    */
    const props = r.response
      ? { message: r.response.data.message, code: r.response.status }
      : { message: r.message, code: 500 };
    const error = Error(props.message);
    error.code = props.code;
    return Promise.reject(error);
  }
}
